tests/python/test_slowmo_fsdp.py

```diff
@@ -0,0 +1,378 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import copy
+import sys
+import tempfile
+from typing import Optional
+
+import torch
+import torch.nn as nn
+from torch import distributed as dist
+from torch.distributed.algorithms.model_averaging import averagers
+from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+from torch.distributed.fsdp.fully_sharded_data_parallel import ShardingStrategy
+from torch.testing._internal.common_distributed import skip_if_lt_x_gpu
+from torch.testing._internal.common_fsdp import FSDPTest
+from torch.testing._internal.common_utils import (
+    instantiate_parametrized_tests,
+    parametrize,
+    run_tests,
+)
+
+from torchdistx.slow_momentum import slow_momentum_comm, slow_momentum_optimizer
+
+if not dist.is_available():
+    print("Distributed not available, skipping tests", file=sys.stderr)
+    sys.exit(0)
+
+
+class Net(nn.Module):
+    def __init__(self, has_wrapping, sharding_strategy):
+        # to ensure determinism
+        torch.manual_seed(0)
+        torch.cuda.manual_seed(0)
+        super().__init__()
+
+        self.linear1 = nn.Linear(8, 16)
+        self.linear2 = nn.Linear(16, 8)
+        self.out = nn.Linear(8, 4)
+
+        fsdp_kwargs = {
+            "device_id": torch.cuda.current_device(),
+            "sharding_strategy": sharding_strategy,
+        }
+
+        self.net = self._maybe_wrap_fsdp(
+            nn.Sequential(
+                self._maybe_wrap_fsdp(
+                    self.linear1, has_wrapping=has_wrapping, **fsdp_kwargs
+                ),
+                nn.ReLU(),
+                self.linear2,
+                nn.ReLU(),
+                self.out,
+            ),
+            has_wrapping=has_wrapping,
+            **fsdp_kwargs
+        )
+
+    def forward(self, x):
+        return self.net(x)
+
+    def _maybe_wrap_fsdp(self, module, has_wrapping, **kwargs):
+        return module if not has_wrapping else FSDP(module, **kwargs)
+
+
+class TestCommunicationHooks(FSDPTest):
+    def _init_fsdp(self, sharding_strategy, net=None):
+        torch.manual_seed(0)
+        torch.cuda.manual_seed(0)
+        net = net if net is not None else torch.nn.Linear(1, 5, bias=False)
+        return FSDP(
+            net,
+            device_id=torch.cuda.current_device(),
+            sharding_strategy=sharding_strategy,
+        )
+
+    def _train_step(self, inpt, net, optim):
+        optim.zero_grad()
+        loss = net(inpt).sum()
+        loss.backward()
+        optim.step()
+
+    def _get_averager(self, period):
```

Comment:
nit: maybe rename this and `_get_slomo_optimizer()` below to start with `_init...` instead of `_get...` since you are constructing these objects for the first time

also `slomo` -> `slowmo`